sw/device/lib/runtime/otbn.c

```diff
@@ -4,16 +4,91 @@
 
 #include "sw/device/lib/runtime/otbn.h"
 
+#include "sw/device/lib/base/freestanding/limits.h"
 #include "sw/device/lib/base/mmio.h"
 #include "sw/device/lib/dif/dif_base.h"
 #include "sw/device/lib/dif/dif_otbn.h"
+#include "sw/device/lib/irq.h"
+#include "sw/device/lib/runtime/ibex.h"
 #include "sw/device/lib/runtime/log.h"
+#include "sw/device/lib/testing/check.h"
+#include "sw/device/lib/testing/rv_plic_testutils.h"
+
+#include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"
 
 /**
  * Data width of big number subset, in bytes.
  */
 const int kOtbnWlenBytes = 256 / 8;
 
+static const uint32_t kPlicTarget = kTopEarlgreyPlicTargetIbex0;
+
+static otbn_t *otbn_ctx;
+
+/**
+ * External interrupt handler.
+ */
+void ottf_external_isr(void) {
+  dif_rv_plic_irq_id_t irq_id;
+  CHECK_DIF_OK(dif_rv_plic_irq_claim(&otbn_ctx->plic, kPlicTarget, &irq_id));
+
+  top_earlgrey_plic_peripheral_t peri = (top_earlgrey_plic_peripheral_t)
+      top_earlgrey_plic_interrupt_for_peripheral[irq_id];
+
+  CHECK(peri == kTopEarlgreyPlicPeripheralOtbn,
+        "Interrupt from incorrect peripheral: (exp: %d, obs: %s)",
+        kTopEarlgreyPlicPeripheralOtbn, peri);
+
+  dif_otbn_irq_t irq = (dif_otbn_irq_t)(
+      irq_id - (dif_rv_plic_irq_id_t)kTopEarlgreyPlicIrqIdOtbnDone);
+
+  // Check this is the interrupt we expected.
+  CHECK(irq_id == kTopEarlgreyPlicIrqIdOtbnDone);
+
+  // Otbn_finished should currently be false (we're supposed to clear it
+  // before starting OTBN).
+  CHECK(!otbn_ctx->finished);
+
+  otbn_ctx->finished = (irq_id == kTopEarlgreyPlicIrqIdOtbnDone);
+
+  CHECK_DIF_OK(dif_otbn_irq_acknowledge(&otbn_ctx->dif, irq));
+
+  // Complete the IRQ by writing the IRQ source to the Ibex specific CC.
+  // register.
+  CHECK_DIF_OK(dif_rv_plic_irq_complete(&otbn_ctx->plic, kPlicTarget, irq_id));
+}
+
+otbn_result_t otbn_busy_wait_for_done_irq(otbn_t *ctx) {
+  // Clear the otbn_finished flag: we'll set it in the interrupt handler when
+  // we see the Done interrupt fire.
+  otbn_ctx = ctx;
+  ctx->finished = false;
+  // Enable Done interrupt.
+  CHECK_DIF_OK(
+      dif_otbn_irq_set_enabled(&ctx->dif, kDifOtbnIrqDone, kDifToggleEnabled));
+
+  // At this point, OTBN should be running. Wait for an interrupt that says
+  // it's done.
+  for (;;) {
+    // This looks a bit odd, but is needed to avoid a race condition where the
+    // OTBN interrupt comes in after we load the otbn_finished flag but before
+    // we run the WFI instruction. The trick is that WFI returns when an
+    // interrupt comes in even if interrupts are globally disabled, which means
+    // that the WFI can actually sit *inside* the critical section.
+    irq_global_ctrl(false);
+    if (ctx->finished)
+      break;
+    wait_for_interrupt();
```

Comment:
This affects the performance measurement as the `mcycles` does not count during the time `wfi` is executed until the interrupt is fired.